src/components/PrintModal/WatermarkModal/WatermarkModal.js

```diff
@@ -0,0 +1,429 @@
+  handleWatermarkRenderFxn;
+
+  static propTypes = {
+    isVisible: PropTypes.bool,
+    pageIndexToView: PropTypes.number,
+    modalClosed: PropTypes.func,
+    formSubmitted: PropTypes.func,
+    t: PropTypes.func.isRequired,
+  }
+
+  constructor(props) {
+    super(props);
+    const locationSettings = this.initializeLocationSettings();
+    this.state = {
+      isColorPaletteVisible: false,
+      locationSettings,
+      previousLocationSettings: locationSettings,
+    };
+    this.canvasContainerRef = React.createRef();
+
+    this.handleWatermarkRenderFxn = () => {
+      if (this.props.isVisible) {
+        this.setState({
+          locationSettings: this.state.previousLocationSettings,
+        }, () => {
+          this.addWatermarks();
+        });
+      } else {
+        this.removeAllWatermarks();
+      }
+    };
+  }
+
+  componentDidMount() {
+    if (this.props.isVisible !== undefined) {
+      core.addEventListener('documentLoaded', this.handleWatermarkRenderFxn);
+    }
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.isVisible !== prevProps.isVisible) {
+      this.setState({
+        isColorPaletteVisible: false,
+      }, () => this.handleWatermarkRenderFxn());
+    }
+  }
+
+  addWatermarks() {
+    const watermarkOptions = this.createWatermarks();
+
+    core.setWatermark(watermarkOptions);
+
+    const pageHeight = core.getPageHeight(this.props.pageIndexToView);
+    const pageWidth = core.getPageWidth(this.props.pageIndexToView);
+
+    const desiredZoomForWidth = DESIRED_WIDTH / pageWidth;
+    const desiredZoomForHeight = DESIRED_HEIGHT / pageHeight;
+
+    const desiredZoom = Math.min(desiredZoomForHeight, desiredZoomForWidth);
+
+    core.getDocument().loadCanvasAsync({
+      pageIndex: this.props.pageIndexToView,
+      zoom: desiredZoom,
+      drawComplete: canvas => {
+        const nodes = this.canvasContainerRef.current.childNodes;
+        if (nodes && nodes.length > 0) {
+          this.canvasContainerRef.current.removeChild(nodes[0]);
+        }
+        canvas.style.border = '1px solid black';
+        canvas.style.height = this.canvasContainerRef.current.style.height;
+        this.canvasContainerRef.current.appendChild(canvas);
+      },
+    });
+
+    // Note: do not update and refresh the doc else it may affect other docs as well
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  constructWatermarkOption(value) {
+    const watermarkOption = {
+      fontSize: value.fontSize,
+      fontFamily: 'sans-serif',
+      color: value.color.toString(),
+      opacity: value.opacity,
+      text: value.text,
+    };
+
+    return watermarkOption;
+  }
+
+  createWatermarks() {
+    let watermarks = {};
+
+    Object.keys(WATERMARK_LOCATIONS).forEach(key => {
+      const temp = this.constructWatermarkOption(this.state.locationSettings[key]);
+
+      switch (WATERMARK_LOCATIONS[key]) {
+        case WATERMARK_LOCATIONS.CENTER:
+          watermarks = {
```

Comment:
Is there a reason this needs to be a new object reference on every iteration of the forEach? Can we just do `watermarks.diagonal = temp;` instead? Or even create a map between WATERMARK_LOCATIONS and the names of the properties on the API?

Reply:
I don't have a strong reason why I keep creating a new object reference on each iteration of the for each. Its probably something I thought of at the spur of that moment when implementing.

I like your suggestion of creating a map between WATERMARK_LOCATIONS and the names of the properties on the API. It cuts down on the huge switch case block that you see. I have changed the code to reflect that.